File: sdk/c/src/providers/temp_mail_io.c
/**
 * temp-mail.io -- https://temp-mail.io
 * REST API: https://api.internal.temp-mail.io/api/v3
 */
#include "tempmail_internal.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define TMIO_BASE "https://api.internal.temp-mail.io/api/v3"
#define TMIO_PAGE "https://temp-mail.io/en"

static char g_tmio_cors_header[128] = "";

static void tmio_try_cache_cors_header(const char *body) {
    const char *p = body ? strstr(body, "mobileTestingHeader") : NULL;
    if (!p) return;

    p = strchr(p, ':');
    if (!p) return;
    p++;
    while (*p == ' ' || *p == '\t' || *p == '\r' || *p == '\n') p++;
    if (*p != '"') return;

    const char *start = p + 1;
    const char *end = strchr(start, '"');
    if (!end || end <= start) return;

    size_t len = (size_t)(end - start);
    if (len >= sizeof(g_tmio_cors_header)) {
        len = sizeof(g_tmio_cors_header) - 1;
    }
    memcpy(g_tmio_cors_header, start, len);
    g_tmio_cors_header[len] = '\0';
}

static const char *tmio_cors_header(void) {
    if (g_tmio_cors_header[0]) {
        return g_tmio_cors_header;
    }

    const char *headers[] = {
        "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/144.0.0.0 Safari/537.36",
        NULL
    };
    tm_http_response_t *resp = tm_http_request(TM_HTTP_GET, TMIO_PAGE, headers, NULL, 15);
    if (resp && resp->body) {
        tmio_try_cache_cors_header(resp->body);
    }
    tm_http_response_free(resp);

    if (!g_tmio_cors_header[0]) {
        strcpy(g_tmio_cors_header, "1");
    }
    return g_tmio_cors_header;
}

static void tmio_headers(const char **headers, char *cors_line, size_t cors_line_cap) {
    snprintf(cors_line, cors_line_cap, "X-CORS-Header: %s", tmio_cors_header());
    headers[0] = "Content-Type: application/json";
    headers[1] = "Application-Name: web";
    headers[2] = "Application-Version: 4.0.0";
    headers[3] = cors_line;
    headers[4] = "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/144.0.0.0 Safari/537.36 Edg/144.0.0.0";
    headers[5] = "Origin: https://temp-mail.io";
    headers[6] = "Referer: https://temp-mail.io/";
    headers[7] = NULL;
}

tm_email_info_t* tm_provider_temp_mail_io_generate(void) {
    char cors_line[192];
    const char *headers[8];
    tmio_headers(headers, cors_line, sizeof(cors_line));

    const char *body = "{\"min_name_length\":10,\"max_name_length\":10}";
    tm_http_response_t *resp = tm_http_request(TM_HTTP_POST, TMIO_BASE "/email/new", headers, body, 15);
    if (!resp || resp->status < 200 || resp->status >= 300) {
        tm_http_response_free(resp);
        return NULL;
    }

    cJSON *root = cJSON_Parse(resp->body);
    tm_http_response_free(resp);
    if (!root) return NULL;

    const char *email = TM_JSON_STR(cJSON_GetObjectItemCaseSensitive(root, "email"), "");
    const char *token = TM_JSON_STR(cJSON_GetObjectItemCaseSensitive(root, "token"), "");
    if (!email[0] || !token[0]) {
        cJSON_Delete(root);
        return NULL;
    }

    tm_email_info_t *info = tm_email_info_new();
    if (!info) {
        cJSON_Delete(root);
        return NULL;
    }
    info->channel = CHANNEL_TEMP_MAIL_IO;
    info->email = tm_strdup(email);
    info->token = tm_strdup(token);

    cJSON_Delete(root);
    return info;
}

tm_email_t* tm_provider_temp_mail_io_get_emails(const char *email, int *count) {
    *count = -1;
    if (!email || !email[0]) return NULL;

    size_t url_len = strlen(TMIO_BASE) + strlen("/email//messages") + strlen(email) + 1;
    char *url = (char*)malloc(url_len);
    if (!url) return NULL;
    snprintf(url, url_len, "%s/email/%s/messages", TMIO_BASE, email);

    char cors_line[192];
    const char *headers[8];
    tmio_headers(headers, cors_line, sizeof(cors_line));

    tm_http_response_t *resp = tm_http_request(TM_HTTP_GET, url, headers, NULL, 15);
    free(url);
    if (!resp || resp->status < 200 || resp->status >= 300) {
        tm_http_response_free(resp);
        return NULL;
    }

    cJSON *root = cJSON_Parse(resp->body);
    tm_http_response_free(resp);
    if (!root) return NULL;

    if (!cJSON_IsArray(root)) {
        *count = 0;
        cJSON_Delete(root);
        return NULL;
    }

    int n = cJSON_GetArraySize(root);
    *count = n;
    if (n == 0) {
        cJSON_Delete(root);
        return NULL;
    }

    tm_email_t *emails = tm_emails_new(n);
    if (!emails) {
        *count = -1;
        cJSON_Delete(root);
        return NULL;
    }

    for (int i = 0; i < n; i++) {
        cJSON *raw = cJSON_GetArrayItem(root, i);
        emails[i] = tm_normalize_email(raw, email);
    }

    cJSON_Delete(root);
    return emails;
}

File: sdk/csharp/src/Providers/TempMailIo.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Text.Json.Nodes;

namespace XxxXTeam.TempMail.Providers;

/// <summary>temp-mail.io 渠道（api.internal.temp-mail.io/api/v3）。</summary>
public static class TempMailIo
{
    private const string BaseUrl = "https://api.internal.temp-mail.io/api/v3";

    private static Dictionary<string, string> BuildHeaders() => new()
    {
        ["Application-Name"] = "web",
        ["Application-Version"] = "4.0.0",
        ["X-CORS-Header"] = "1",
        ["User-Agent"] = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/144.0.0.0 Safari/537.36 Edg/144.0.0.0",
        ["Origin"] = "https://temp-mail.io",
        ["Referer"] = "https://temp-mail.io/",
    };

    public static EmailInfo Generate()
    {
        var payload = Json.Serialize(new Dictionary<string, object?> { ["min_name_length"] = 10, ["max_name_length"] = 10 });
        var resp = Http.Post($"{BaseUrl}/email/new", payload, "application/json", BuildHeaders());
        resp.EnsureSuccess();
        var data = Json.Parse(resp.Body);
        var email = Json.Str(data, "email");
        var token = Json.Str(data, "token");
        if (string.IsNullOrEmpty(email) || string.IsNullOrEmpty(token))
            throw new Exception("temp-mail-io: invalid generate response");
        return new EmailInfo("temp-mail-io", email, token);
    }

    public static List<Email> GetEmails(string email)
    {
        var resp = Http.Get($"{BaseUrl}/email/{WebUtility.UrlEncode(email)}/messages", BuildHeaders());
        resp.EnsureSuccess();
        var rows = Json.Parse(resp.Body) as JsonArray;
        var result = new List<Email>();
        if (rows is null) return result;
        foreach (var raw in rows)
            if (raw is JsonObject) result.Add(Normalize.NormalizeEmail(Json.ToDict(raw), email));
        return result;
    }
}

/// <summary>mail.cx 匿名 Web API 渠道（mail.cx/v1）。ddker.com 复用。</summary>
public static class MailCx
{
    private const string BaseUrl = "https://mail.cx";
    private static readonly Random Rand = new();

    private static string RandomString(int len)
    {
        const string chars = "abcdefghijklmnopqrstuvwxyz0123456789";
        var sb = new System.Text.StringBuilder();
        for (var i = 0; i < len; i++) sb.Append(chars[Rand.Next(chars.Length)]);
        return sb.ToString();
    }

    private static Dictionary<string, string> BuildHeaders(string clientId) => new()
    {
        ["Accept"] = "application/json",
        ["Origin"] = BaseUrl,
        ["Referer"] = BaseUrl + "/",
        ["User-Agent"] = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0",
        ["X-Client-ID"] = clientId,
    };

    public static EmailInfo Generate(string channel, string? preferredDomain)
    {
        var clientId = "tempmail-sdk-" + RandomString(16);
        var resp = Http.Get($"{BaseUrl}/v1/config", BuildHeaders(clientId));
        resp.EnsureSuccess();
        var config = Json.Parse(resp.Body) as JsonObject;

        var domain = PickDomain(config, preferredDomain);
        return new EmailInfo(channel, $"{RandomString(12)}@{domain}", clientId);
    }

    private static string PickDomain(JsonObject? config, string? preferred)
    {
        var items = config?["system_domains"] as JsonArray;
        var domains = new List<string>();
        if (items is not null)
            foreach (var it in items)
                if (it is JsonObject o)
                {
                    var d = (o["domain"]?.GetValue<string>() ?? "").Trim();
                    if (d.Length > 0) domains.Add(d);
                }
        if (domains.Count == 0) throw new Exception("mail-cx: no system domains");

        var want = (preferred ?? "").Trim().TrimStart('@').ToLowerInvariant();
        if (want.Length > 0)
            foreach (var d in domains)
                if (d.ToLowerInvariant() == want) return d;

        if (items is not null)
            foreach (var it in items)
                if (it is JsonObject o && (o["default"]?.GetValue<bool>() ?? false))
                {
                    var d = (o["domain"]?.GetValue<string>() ?? "").Trim();
                    if (d.Length > 0) return d;
                }
        return domains[0];
    }

    public static List<Email> GetEmails(string? clientId, string email)
    {
        var resp = Http.Get($"{BaseUrl}/v1/inbox/{WebUtility.UrlEncode(email)}", BuildHeaders(clientId ?? ""));
        if (resp.StatusCode == 204) return new List<Email>();
        resp.EnsureSuccess();
        var rows = (Json.Parse(resp.Body) as JsonObject)?["emails"] as JsonArray;
        var result = new List<Email>();
        if (rows is null) return result;
        foreach (var row in rows)
        {
            if (row is not JsonObject ro) { result.Add(Normalize.NormalizeEmail(new Dictionary<string, object?>(), email)); continue; }
            var raw = new Dictionary<string, object?>
            {
                ["id"] = Json.Str(ro, "id"),
                ["from"] = FirstNonEmpty(Json.Str(ro, "from_email"), Json.Str(ro, "from_name")),
                ["to"] = email,
                ["subject"] = Json.Str(ro, "subject"),
                ["text"] = Json.Str(ro, "preview_text"),
                ["created_at"] = Json.Str(ro, "created_at"),
            };
            var id = Json.Str(ro, "id");
            if (!string.IsNullOrEmpty(id))
            {
                try
                {
                    var dr = Http.Get($"{BaseUrl}/v1/email/{WebUtility.UrlEncode(id)}", BuildHeaders(clientId ?? ""));
                    if (dr.Ok && Json.Parse(dr.Body) is JsonObject det)
                        raw = new Dictionary<string, object?>
                        {
                            ["id"] = Json.Str(det, "id"),
                            ["from"] = FirstNonEmpty(Json.Str(det, "from_email"), Json.Str(det, "from_name")),
                            ["to"] = email,
                            ["subject"] = Json.Str(det, "subject"),
                            ["text"] = FirstNonEmpty(Json.Str(det, "text_body"), Json.Str(det, "preview_text")),
                            ["html"] = Json.Str(det, "html_body"),
                            ["created_at"] = Json.Str(det, "created_at"),
                        };
                }
                catch { /* 回退列表摘要 */ }
            }
            result.Add(Normalize.NormalizeEmail(raw, email));
        }
        return result;
    }

    private static string FirstNonEmpty(params string[] values)
    {
        foreach (var v in values) if (!string.IsNullOrWhiteSpace(v)) return v.Trim();
        return "";
    }
}
